File: src/main/java/com/github/svegon/capi/event/render/tooltip/BarrelTooltipCallback.java
package com.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.block.BarrelBlock;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.world.BlockView;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@FunctionalInterface
public interface BarrelTooltipCallback {
    Event<BarrelTooltipCallback> EVENT = EventFactory.createArrayBacked(BarrelTooltipCallback.class,
            (block,stack, world, tooltip, options) -> {}, listeners -> (block,stack, world, tooltip, context) -> {
        for (BarrelTooltipCallback listener : listeners) {
            listener.appendBarrelTooltip(block, stack, world, tooltip, context);
        }
    });

    void appendBarrelTooltip(BarrelBlock block, ItemStack stack, @Nullable BlockView world,
                             List<Text> tooltip, TooltipContext options);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/BarrelTooltipCallback.kt
package io.github.svegon.capi.event.render.tooltip

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.item.TooltipContext
import net.minecraft.text.Text
import net.minecraft.world.BlockView

fun interface BarrelTooltipCallback {
    fun appendBarrelTooltip(
        block: BarrelBlock?, stack: ItemStack?, world: BlockView?,
        tooltip: List<Text?>?, options: TooltipContext?
    )

    companion object {
        @JvmField
        val EVENT: Event<BarrelTooltipCallback> = EventFactory.createArrayBacked<BarrelTooltipCallback>(
            BarrelTooltipCallback::class.java,
            BarrelTooltipCallback { block: BarrelBlock?, stack: ItemStack?, world: BlockView?, tooltip: List<Text?>?, options: TooltipContext? -> }) { listeners: Array<BarrelTooltipCallback> ->
            BarrelTooltipCallback { block: BarrelBlock?, stack: ItemStack?, world: BlockView?, tooltip: List<Text?>?, context: TooltipContext? ->
                for (listener in listeners) {
                    listener.appendBarrelTooltip(block, stack, world, tooltip, context)
                }
            }
        }
    }
}
